chain/solana/src/data_source.rs

```diff
@@ -0,0 +1,406 @@
+use base58::ToBase58;
+use graph::blockchain::{Block, TriggerWithHandler};
+use graph::components::store::StoredDynamicDataSource;
+use graph::data::subgraph::DataSourceContext;
+use graph::prelude::SubgraphManifestValidationError;
+use graph::{
+    anyhow::{anyhow, Error},
+    blockchain::{self, Blockchain},
+    prelude::{
+        async_trait, info, BlockNumber, CheapClone, DataSourceTemplateInfo, Deserialize, Link,
+        LinkResolver, Logger,
+    },
+    semver,
+};
+use std::collections::BTreeMap;
+use std::{convert::TryFrom, sync::Arc};
+
+use crate::chain::Chain;
+use crate::trigger::SolanaTrigger;
+
+pub const SOLANA_KIND: &str = "solana";
+
+/// Runtime representation of a data source.
+#[derive(Clone, Debug)]
+pub struct DataSource {
+    pub kind: String,
+    pub network: Option<String>,
+    pub name: String,
+    pub(crate) source: Source,
+    pub mapping: Mapping,
+    pub context: Arc<Option<DataSourceContext>>,
+    pub creation_block: Option<BlockNumber>,
+}
+
+impl blockchain::DataSource<Chain> for DataSource {
+    fn address(&self) -> Option<&[u8]> {
+        self.source.program_id.as_ref().map(String::as_bytes)
+    }
+
+    fn start_block(&self) -> BlockNumber {
+        self.source.start_block
+    }
+
+    fn name(&self) -> &str {
+        &self.name
+    }
+
+    fn kind(&self) -> &str {
+        &self.kind
+    }
+
+    fn network(&self) -> Option<&str> {
+        self.network.as_ref().map(|s| s.as_str())
+    }
+
+    fn context(&self) -> Arc<Option<DataSourceContext>> {
+        self.context.cheap_clone()
+    }
+
+    fn creation_block(&self) -> Option<BlockNumber> {
+        self.creation_block
+    }
+
+    fn api_version(&self) -> semver::Version {
+        self.mapping.api_version.clone()
+    }
+
+    fn runtime(&self) -> &[u8] {
+        self.mapping.runtime.as_ref()
+    }
+
+    fn match_and_decode(
+        &self,
+        trigger: &<Chain as Blockchain>::TriggerData,
+        block: Arc<<Chain as Blockchain>::Block>,
+        _logger: &Logger,
+    ) -> Result<Option<TriggerWithHandler<Chain>>, Error> {
+        if self.source.start_block > block.number() {
+            return Ok(None);
+        }
+
+        let handler = match trigger {
+            // A block trigger matches if a block handler is present.
+            SolanaTrigger::Block(_) => match self.handler_for_block() {
+                Some(handler) => &handler.handler,
+                None => return Ok(None),
+            },
+
+            SolanaTrigger::Instruction(instruction_with_block) => {
+                let pid = &instruction_with_block.instruction.program_id;
```

Comment:
`pid` when I read it later does not directly helps understanding, don't think it's a good acronym in this context here.